### PR TITLE
Attach weights to DeltaLakeSplits

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Attach weights to DeltaLakeSplits.

## Documentation

(x) Sufficient documentation is included in this PR.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Delta Lake
* Use split weights to improve scheduling and query performance on tables with many small files. ({issue}`12755`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplit.java [hunk 1]
```diff
@@ -46,6 +47,7 @@
     private final List<HostAddress> addresses;
     private final TupleDomain<DeltaLakeColumnHandle> statisticsPredicate;
     private final Map<String, Optional<String>> partitionKeys;
+    private final SplitWeight splitWeight;
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplit.java:50]: Nit: match the order from the constr, put this between addresses and statisticsPredicate

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java [hunk 2]
```diff
@@ -235,6 +238,7 @@ private List<DeltaLakeSplit> splitsForFile(
                     fileSize,
                     addFileEntry.getModificationTime(),
                     ImmutableList.of(),
+                    SplitWeight.fromProportion(Math.min(Math.max((double) fileSize, minimumAssignedSplitWeight), 1.0)),
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java:241]: This is always going to be `1.0`. I think for non-splittable queries it's okay to just always set this as `SplitWeight.standard()`. This is only used for update/delete queries anyway

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java [hunk 3]
```diff
@@ -259,6 +263,7 @@ private List<DeltaLakeSplit> splitsForFile(
                     fileSize,
                     addFileEntry.getModificationTime(),
                     ImmutableList.of(),
+                    SplitWeight.fromProportion(Math.min(Math.max((double) fileSize / splitSize, minimumAssignedSplitWeight), 1.0)),
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java:266]: `fileSize / splitSize` I think what we want here is the `splitSize` divided by the intended split size. 

Tweaked from the code above:
```
            long maxSplitSize;
            if (remainingInitialSplits.get() > 0 && remainingInitialSplits.getAndDecrement() > 0) {
                maxSplitSize = getMaxInitialSplitSize(session).toBytes();
            }
            else {
                maxSplitSize = getMaxSplitSize(session).toBytes();
            }
            long splitSize = Math.min(maxSplitSize, fileSize - currentOffset);

            SplitWeight weight = SplitWeight.fromProportion(Math.min(Math.max((double) splitSize / maxSplitSize, minimumAssignedSplitWeight), 1.0))
```

**Review by alexjo2144 — CHANGES_REQUESTED** [review]

**Review by alexjo2144 — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java [hunk 4]
```diff
@@ -259,6 +262,7 @@ private List<DeltaLakeSplit> splitsForFile(
                     fileSize,
                     addFileEntry.getModificationTime(),
                     ImmutableList.of(),
+                    SplitWeight.fromProportion(Math.min(Math.max((double) splitSize / maxSplitSize, minimumAssignedSplitWeight), 1.0)),
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java:265]: `Math.min( ..., 1)` is redundant per my understanding.

nit: static import min & max

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java (hunk 4 above)
Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java:265]: > `Math.min( ..., 1)` is redundant per my understanding.

ie if we somehow end up creating an oversized split, it's fine to let the engine know about that.

**Comment by findepi** [comment]
@ebyhr you can merge this as is and i will update https://github.com/trinodb/trino/pull/12832 to cover Delta too.

**Comment by ebyhr** [comment]
Let me merge this as-is. Thanks for your follow-up PR.

**Pull request merged**